homeassistant/components/mqtt/siren.py

```diff
@@ -0,0 +1,374 @@
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_OPTIMISTIC, default=DEFAULT_OPTIMISTIC): cv.boolean,
+        vol.Optional(CONF_PAYLOAD_OFF, default=DEFAULT_PAYLOAD_OFF): cv.string,
+        vol.Optional(CONF_PAYLOAD_ON, default=DEFAULT_PAYLOAD_ON): cv.string,
+        vol.Optional(CONF_STATE_OFF): cv.string,
+        vol.Optional(CONF_STATE_ON): cv.string,
+        vol.Optional(CONF_STATE_VALUE_TEMPLATE): cv.template,
+        vol.Optional(CONF_SUPPORT_DURATION, default=True): cv.boolean,
+        vol.Optional(CONF_SUPPORT_VOLUME_SET, default=True): cv.boolean,
+    },
+).extend(MQTT_ENTITY_COMMON_SCHEMA.schema)
+
+DISCOVERY_SCHEMA = vol.All(PLATFORM_SCHEMA.extend({}, extra=vol.REMOVE_EXTRA))
+
+MQTT_SIREN_ATTRIBUTES_BLOCKED = frozenset(
+    {
+        ATTR_AVAILABLE_TONES,
+        ATTR_DURATION,
+        ATTR_TONE,
+        ATTR_VOLUME_LEVEL,
+    }
+)
+
+SUPPORTED_BASE = SUPPORT_TURN_OFF | SUPPORT_TURN_ON
+
+SUPPORTED_ATTRIBUTES = {
+    ATTR_DURATION: SUPPORT_DURATION,
+    ATTR_TONE: SUPPORT_TONES,
+    ATTR_VOLUME_LEVEL: SUPPORT_VOLUME_SET,
+}
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    async_add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
+    """Set up MQTT siren through configuration.yaml."""
+    await async_setup_reload_service(hass, DOMAIN, PLATFORMS)
+    await _async_setup_entity(hass, async_add_entities, config)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up MQTT siren dynamically through MQTT discovery."""
+    setup = functools.partial(
+        _async_setup_entity, hass, async_add_entities, config_entry=config_entry
+    )
+    await async_setup_entry_helper(hass, siren.DOMAIN, setup, DISCOVERY_SCHEMA)
+
+
+async def _async_setup_entity(
+    hass, async_add_entities, config, config_entry=None, discovery_data=None
+):
+    """Set up the MQTT siren."""
+    async_add_entities([MqttSiren(hass, config, config_entry, discovery_data)])
+
+
+class MqttSiren(MqttEntity, SirenEntity):
+    """Representation of a siren that can be controlled using MQTT."""
+
+    _entity_id_format = ENTITY_ID_FORMAT
+    _attributes_extra_blocked = MQTT_SIREN_ATTRIBUTES_BLOCKED
+
+    def __init__(self, hass, config, config_entry, discovery_data):
+        """Initialize the MQTT siren."""
+        self._attr_name = config[CONF_NAME]
+        self._attr_should_poll = False
+        self._supported_features = SUPPORTED_BASE
+        self._attr_is_on = None
+        self._state_on = None
+        self._state_off = None
+        self._optimistic = None
+
+        self._attr_extra_state_attributes: dict[str, Any] = {}
+
+        self.target = None
+
+        MqttEntity.__init__(self, hass, config, config_entry, discovery_data)
```

Comment:
We could use `super()` here since there's no confusion about which parent init method to target.

Reply:
I'll have a look

Reply:
See https://github.com/home-assistant/core/pull/65291